src/hotspot/share/cds/metaspaceShared.cpp

```diff
@@ -1486,6 +1486,15 @@ char* MetaspaceShared::reserve_address_space_for_archives(FileMapInfo* static_ma
   const size_t total_range_size =
       archive_space_size + gap_size + class_space_size;
 
+  // Test that class space base address plus shift can be decoded by aarch64, when restored.
+  const int precomputed_narrow_klass_shift = ArchiveBuilder::precomputed_narrow_klass_shift();
+  if (!CompressedKlassPointers::check_klass_decode_mode(base_address, precomputed_narrow_klass_shift,
+                                                        total_range_size)) {
+    log_info(cds)("CDS initialization: Cannot use SharedBaseAddress " PTR_FORMAT " with precomputed shift %d.",
+                  p2i(base_address), precomputed_narrow_klass_shift);
+    use_archive_base_addr = false;
+  }
```

Comment:
If this is a log-error or fatal, then maybe we should give an error for an invalid SharedBaseAddress, as with Thomas's suggested patch.  We generally try to get SharedBaseAddress and that fails, let the OS code pick one.  If we want to give an error for invalid SharedBaseAddress, I could fix the test that fails to not try for an invalid address.  But the validity of the address is only for aarch64 so I don't know if we want to do this.

Reply:
`base_address` here is not from user input. It is something produced by the JVM after validity checks. So if we have an error it, we JVM has done something wrong. I think a fatal() would be apropriate.

In contract, SharedBaseAddress can be specified by the user, and we usually do not use fatal() to report such errors.

Reply:
SharedBaseAddress comes through here also.  It was valid at one point but UseCompactObjectHeaders forces the shift to be 10 so on using the shared archive, it might not be correct.
My observation is that SharedBaseAddress is ignored most of the time, which is why this failure was intermittent.

Reply:
I added a comment why the validity check in -Xshare:dump doesn't actually check this and doesn't really have the full reservation size so it can check it.